File: tests/runners/rust/tests/test_03_nested_namespaces.rs
// Test Case 03: Nested Namespaces
// Tests deeply nested namespace structures

use std::io::Cursor;

use polygen_test::schema::app::app_data::app_data_models::User;
use polygen_test::schema::app::app_data::app_data_enums::Permission;
use polygen_test::schema::app::app_services::UserService;
use polygen_test::schema::util::Config;
use polygen_test::schema_loaders::BinaryIO;

fn main() {
    println!("=== Test Case 03: Nested Namespaces ===");

    test_deeply_nested_table();
    test_nested_enum();
    test_cross_namespace_reference();
    test_separate_namespace();
    test_binary_nested();

    println!("=== All tests passed! ===");
}

fn test_deeply_nested_table() {
    println!("  Testing deeply nested table (app.data.models.User)...");

    let user = User {
        id: 1,
        username: "testuser".to_string(),
    };

    assert_eq!(user.id, 1);
    assert_eq!(user.username, "testuser");

    println!("    PASS");
}

fn test_nested_enum() {
    println!("  Testing nested enum (app.data.enums.Permission)...");

    let perm = Permission::Admin;
    assert_eq!(perm as i32, 3);

    let perm = Permission::Read;
    assert_eq!(perm as i32, 1);

    println!("    PASS");
}

fn test_cross_namespace_reference() {
    println!("  Testing cross-namespace reference (UserService)...");

    let service = UserService {
        id: 1,
        target_user_id: 42,
        permission: Permission::Write,
    };

    assert_eq!(service.id, 1);
    assert_eq!(service.target_user_id, 42);
    assert_eq!(service.permission, Permission::Write);

    println!("    PASS");
}

fn test_separate_namespace() {
    println!("  Testing separate namespace (util.Config)...");

    let config = Config {
        key: "debug_mode".to_string(),
        value: "true".to_string(),
    };

    assert_eq!(config.key, "debug_mode");
    assert_eq!(config.value, "true");

    println!("    PASS");
}

fn test_binary_nested() {
    println!("  Testing binary serialization with nested namespaces...");

    let original = UserService {
        id: 999,
        target_user_id: 42,
        permission: Permission::Admin,
    };

    // Serialize
    let mut buffer = Vec::new();
    original.write_binary(&mut buffer).unwrap();

    // Deserialize
    let mut cursor = Cursor::new(&buffer);
    let loaded = UserService::read_binary(&mut cursor).unwrap();

    assert_eq!(loaded.id, original.id);
    assert_eq!(loaded.target_user_id, original.target_user_id);
    assert_eq!(loaded.permission, original.permission);

    println!("    PASS (serialized {} bytes)", buffer.len());
}

File: tests/runners/typescript/tests/test_03_nested_namespaces.ts
// Test Case 03: Nested Namespaces
// Tests deeply nested namespace structures

import { App, Util } from '../generated/03_nested_namespaces/typescript/schema';
import * as ZodSchemas from '../generated/03_nested_namespaces/typescript/schema.schema';

// Test deeply nested table
function testDeeplyNestedTable(): void {
    console.log("  Testing deeply nested table (App.AppData.AppDataModels.User)...");

    const user: App.AppData.AppDataModels.User = {
        id: 1,
        username: "testuser",
    };

    console.assert(user.id === 1, "id should be 1");
    console.assert(user.username === "testuser", "username should be testuser");

    console.log("    PASS");
}

// Test nested enum
function testNestedEnum(): void {
    console.log("  Testing nested enum (App.AppData.AppDataEnums.Permission)...");

    const read = App.AppData.AppDataEnums.Permission.Read;
    console.assert(read === 1, "Read should be 1");

    const write = App.AppData.AppDataEnums.Permission.Write;
    console.assert(write === 2, "Write should be 2");

    const admin = App.AppData.AppDataEnums.Permission.Admin;
    console.assert(admin === 3, "Admin should be 3");

    console.log("    PASS");
}

// Test cross-namespace reference (UserService)
function testCrossNamespaceReference(): void {
    console.log("  Testing cross-namespace reference (UserService)...");

    const service: App.AppServices.UserService = {
        id: 1,
        targetUserId: 100,
        permission: App.AppData.AppDataEnums.Permission.Admin,
    };

    console.assert(service.id === 1, "id should be 1");
    console.assert(service.targetUserId === 100, "targetUserId should be 100");
    console.assert(service.permission === App.AppData.AppDataEnums.Permission.Admin, "permission should be Admin");

    console.log("    PASS");
}

// Test separate namespace (Util.Config)
function testSeparateNamespace(): void {
    console.log("  Testing separate namespace (Util.Config)...");

    const config: Util.Config = {
        key: "app.setting",
        value: "enabled",
    };

    console.assert(config.key === "app.setting", "key should match");
    console.assert(config.value === "enabled", "value should match");

    console.log("    PASS");
}

// Test Zod schemas for nested namespaces
function testNestedZodSchemas(): void {
    console.log("  Testing nested namespace Zod schemas...");

    const userResult = ZodSchemas.App.AppData.AppDataModels.UserSchema.safeParse({
        id: 7,
        username: "deep_user",
    });
    console.assert(userResult.success, "UserSchema should accept valid nested user");

    const serviceResult = ZodSchemas.App.AppServices.UserServiceSchema.safeParse({
        id: 11,
        targetUserId: 7,
        permission: App.AppData.AppDataEnums.Permission.Read,
    });
    console.assert(serviceResult.success, "UserServiceSchema should accept valid cross-namespace enum reference");

    const invalidServiceResult = ZodSchemas.App.AppServices.UserServiceSchema.safeParse({
        id: 12,
        targetUserId: 7,
        permission: 999,
    });
    console.assert(!invalidServiceResult.success, "UserServiceSchema should reject invalid enum value");

    const configResult = ZodSchemas.Util.ConfigSchema.safeParse({
        key: "debug_mode",
        value: "true",
    });
    console.assert(configResult.success, "ConfigSchema should accept separate namespace config");

    console.log("    PASS");
}

// Main
console.log("=== Test Case 03: Nested Namespaces ===");
testDeeplyNestedTable();
testNestedEnum();
testCrossNamespaceReference();
testSeparateNamespace();
testNestedZodSchemas();
console.log("=== All tests passed! ===");
